src/ripple/core/SQLInterface.h

```diff
@@ -0,0 +1,674 @@
+        Application& app,
+        uint256 const& id,
+        boost::optional<ClosedInterval<uint32_t>> const& range,
+        error_code_i& ec) = 0;
+
+    virtual bool
+    checkDBSpace(
+        SQLDatabase& txDb,
+        Config const& config,
+        beast::Journal& j) = 0;
+
+    virtual std::pair<PublicKey, SecretKey>
+    loadNodeIdentity(SQLDatabase& db) = 0;
+
+    virtual void
+    databaseBodyDoPut(
+        SQLDatabase& conn,
+        std::string& data,
+        std::string& path,
+        std::uint64_t fileSize,
+        std::uint64_t& part,
+        std::uint16_t const maxRowSizePad) = 0;
+
+    virtual void
+    databaseBodyFinish(SQLDatabase& conn, std::ofstream& fout) = 0;
+
+    virtual void
+    addValidatorManifest(SQLDatabase& db, std::string const& serialized) = 0;
+
+    virtual void
+    loadPeerReservationTable(
+        SQLDatabase& conn,
+        beast::Journal& j,
+        std::unordered_set<PeerReservation, beast::uhash<>, KeyEqual>&
+            table) = 0;
+
+    virtual void
+    insertPeerReservation(
+        SQLDatabase& conn,
+        PublicKey const& nodeId,
+        std::string const& description) = 0;
+
+    virtual void
+    deletePeerReservation(SQLDatabase& conn, PublicKey const& nodeId) = 0;
+
+    virtual void
+    readArchiveDB(
+        SQLDatabase& db,
+        std::function<void(std::string const&, int)> const& func) = 0;
+
+    virtual void
+    insertArchiveDB(
+        SQLDatabase& db,
+        LedgerIndex shardIndex,
+        std::string const& url) = 0;
+
+    virtual void
+    deleteFromArchiveDB(SQLDatabase& db, LedgerIndex shardIndex) = 0;
+
+    virtual void
+    dropArchiveDB(SQLDatabase& db) = 0;
+
+    virtual int
+    getKBUsedAll(SQLDatabase& db) = 0;
+
+    virtual int
+    getKBUsedDB(SQLDatabase& db) = 0;
+
+    virtual void
+    readPeerFinderDB(
+        SQLDatabase& db,
+        std::function<void(std::string const&, int)> const& func) = 0;
+
+    virtual void
+    savePeerFinderDB(
+        SQLDatabase& db,
+        std::vector<PeerFinder::Store::Entry> const& v) = 0;
+
+protected:
+    static LedgerIndex
+    seqToShardIndex(LedgerIndex seq);
+
+    static LedgerIndex
+    firstLedgerSeq(LedgerIndex shardIndex);
+
+    static LedgerIndex
+    lastLedgerSeq(LedgerIndex shardIndex);
+
+    static LedgerIndex ledgersPerShard_;
+
+private:
+    static std::unordered_map<SQLInterface::DatabaseType, SQLInterface*>
+        type2iface_;
+    static std::unordered_map<SQLDatabase_*, DatabaseIndex> db2ind_;
+    static std::map<LedgerIndex, SQLDatabase_*> txInd2db_;
+    static std::map<LedgerIndex, SQLDatabase_*> lgrInd2db_;
+    static std::mutex maps_mutex_;
+};
+
+class SQLDatabase_
+{
+public:
+    SQLDatabase_(SQLInterface* iface) : iface_(iface)
+    {
+    }
+
+    virtual ~SQLDatabase_()
+    {
+        SQLInterface::removeDatabase(this);
+    }
+
+    SQLInterface*
+    getInterface()
+    {
+        return iface_;
+    }
+
+    SQLInterface* iface_;
```

Comment:
I think the indirection between `SQLInterace` and `SQLDatabase` is confusing. `SQLDatabase` has a `SQLInterface` data member, but a bunch of methods of `SQLInterface` take a `SQLDatabase` as a parameter. I think we should merge these two classes/interfaces